presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneTableScanColumns.java

```diff
@@ -13,34 +13,109 @@
 
 public class PruneTableScanColumns
         extends ProjectOffPushDownRule<TableScanNode>
 {
-    public PruneTableScanColumns()
+    private final Metadata metadata;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PruneTableScanColumns(Metadata metadata, TypeAnalyzer typeAnalyzer)
     {
         super(tableScan());
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
     }
 
     @Override
-    protected Optional<PlanNode> pushDownProjectOff(Context context, TableScanNode tableScanNode, Set<Symbol> referencedOutputs)
+    protected Optional<PlanNode> pushDownProjectOff(Context context, TableScanNode node, Set<Symbol> referencedOutputs)
     {
-        return Optional.of(
-                new TableScanNode(
-                        tableScanNode.getId(),
-                        tableScanNode.getTable(),
-                        filteredCopy(tableScanNode.getOutputSymbols(), referencedOutputs::contains),
-                        filterKeys(tableScanNode.getAssignments(), referencedOutputs::contains),
-                        tableScanNode.getEnforcedConstraint()));
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+
+        return pruneColumns(metadata, typeAnalyzer, types, session, node, referencedOutputs);
+    }
+
+    public static Optional<PlanNode> pruneColumns(Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider types, Session session, TableScanNode node, Set<Symbol> referencedOutputs)
+    {
+        List<Symbol> newOutputs = node.getOutputSymbols().stream()
+                .filter(referencedOutputs::contains)
+                .collect(toImmutableList());
+
+        if (newOutputs.size() == node.getOutputSymbols().size()) {
+            return Optional.empty();
+        }
+
+        List<ConnectorExpression> projections = newOutputs.stream()
+                .map(Symbol::toSymbolReference)
+                .map(expression -> ConnectorExpressionTranslator.translate(
+                        session,
+                        expression,
+                        typeAnalyzer,
+                        types))
+                .map(Optional::get)
+                .collect(toImmutableList());
+
+        TableHandle handle = node.getTable();
+        Optional<ProjectionApplicationResult<TableHandle>> result = metadata.applyProjection(
+                session,
+                handle,
+                projections,
+                newOutputs.stream()
+                        .collect(Collectors.toMap(Symbol::getName, node.getAssignments()::get)));
+
+        Map<Symbol, ColumnHandle> newAssignments;
+        // Attempt to push down the constrained list of columns into the connector.
+        // Bail out if the connector does anything other than limit the list of columns (e.g., if it synthesizes arbitrary expressions)
```

Comment:
No, it's perfectly legal for a connector to produce projections other than simple column references. For instance, it might decide that a column is constant and return a constant expression in its place.

> Otherwise applyProjection is ineffective.

See my comment here https://github.com/prestosql/presto/pull/3305#discussion_r402445247

Reply:
> it's perfectly legal for a connector to produce projections other than simple column references

I get that.

However, we trying here to pushdown limiting identity projection here (no fancy expressions here). Connector returned new table handle and some expressions. In case when we notice that there are some expressions we do not support, we should fail and not to try to return table scan node that has old handle and new assignments (that do not match table handle stored in table scan node) like it is right now.  Is it safe to just ignore what connector returned?

Reply:
We could ask connector to project all ROW's fields.
And connector says "hey, i will give you the ROW itself", and please do projections on the engine side.

I don't know why connector would choose that, but that's supposed to be legal, but cannot be accepted with current code.

@martint i think this `if` deserves a TODO note, along maybe with "merge PruneTableScanColumns with PushProjectionIntoTableScan rules".

Reply:
> I don't know why connector would choose that, but that's supposed to be legal, but cannot be accepted with current code.

Indeed. That's beyond the scope of this change and is handled by PushProjectionIntoTableScan (given the expectation that when pruning columns with rules there's always a projection above the tablescan). As I said above, this rule should probably be replaced by PushProjectionIntoTableScan, but I didn't want to go down that path just yet as it's a more involved process of tracking down all the places where it needs to be replaced, ensuring there are no other unwanted side-effects, etc.

I'll add a TODO, though. That seems fair.